### PR TITLE
Add namespace filter

### PR DESCRIPTION
**Pull request by tom-dudley**
Implementation of: https://github.com/nunit/nunit/issues/1990

I've added what I believe are the required changes on the framework side to add support for the namespace keyword. I'm not sure what the deal with this getting added to the different versions of .NET, but since I believe everything added it compatible, I went ahead with adding to all of them.

nunit-console change dependent on this PR is here: https://github.com/nunit/nunit-console/pull/165

### CODE REVIEW

**Comment by tom-dudley** [comment]
Ah yes, good point! 

Is there going to be any issue with me doing an interactive rebase and force pushing back to here? I see I messed up some csproj files in that second commit.. 

Speaking of, what's the recommended way to do development on the framework and console at the same time? I ended up just creating a solution and pulling in the separate projects but I'm certain there's a better way!

**Comment by tom-dudley** [comment]
I'm a bit confused about the various ExplicitMatchTests by the way. Looks like the tests are disabled in a few of the filters and enabling them makes them fail. Ideally I'd add a test case for the expected behaviour, but I can't tell what that should be!

**Comment by CharliePoole** [comment]
No problem with that. We don't consider any branches as "published" except for master.

Generally, we try to work in the two projects separately, although you can experiment locally as you are doing. We do occasionally have issues of dependency but not here. Your console changes will work differently with a newer framework but it has to handle the older versions as well.

Once the framework changes are merged, the nuget dev package will be available on myget. That's why the console issue is labeled as blocked.

**Comment by ChrisMaddock** [comment]
> I'm a bit confused about the various ExplicitMatchTests by the way. Looks like the tests are disabled in a few of the filters and enabling them makes them fail. Ideally I'd add a test case for the expected behaviour, but I can't tell what that should be!

Not totally sure what you're looking at here @tom-dudley - did this get resolved? If not, can you be more specific? I can't find the disabled tests?

**Comment by tom-dudley** [comment]
Aha, I was about to say I couldn't find them, but here we go:

Note the absence of the [Test] attribute [here](https://github.com/nunit/nunit/blob/757feb7a721ee8ef21b20e76ada61d00834c3bc7/src/NUnitFramework/tests/Internal/Filters/FullNameFilterTests.cs#L63-L64)

Adding the attribute makes the test fail*, hence my confusion.

*It might actually pass on this one, but all the filter tests seem to have the attribute missing for this test. Some of them do fail though.

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        /// </summary>
+        public override bool Match(ITest test)
+        {
+            string containingNamespace = null;
+
+            if (test.TypeInfo != null)
+            {
+                containingNamespace = test.TypeInfo.Namespace;
+            }
+            else if (test is TestSuite || test is TestFixture)
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: Could be simplified to `(test is TestSuite)`

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                if (test.HasChildren)
+                {
+                    containingNamespace = test.Tests[0].TypeInfo.Namespace;
+                }
+                else
+                {
+                    var outerMostClass = test.ClassName.Substring(0, test.ClassName.IndexOf("+"));
+                    containingNamespace = outerMostClass.Substring(0, outerMostClass.LastIndexOf("."));
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: This will crash if `test.ClassName` is null, does not contain `+`, or does not contain `.`.

I don't know why `ClassName` would ever be null, but I just saw it happen. Then against I don't understand why `TypeInfo` would ever be null lol.

Could you add a test for a non-nested class in the global namespace, where the entirety of `ClassName` is something like `Foo`? An empty string should be returned for the namespace.

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+            if (test.TypeInfo != null)
+            {
+                containingNamespace = test.TypeInfo.Namespace;
+            }
+            else if (test is TestSuite || test is TestFixture)
+            {
+                if (test.HasChildren)
+                {
+                    containingNamespace = test.Tests[0].TypeInfo.Namespace;
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: I'm curious, do we have a good understanding of why `test.TypeInfo` is null and why `test.Tests[0].TypeInfo` would not be null?

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs (hunk 2 above)
Comment (tom-dudley) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: @jnm2 I think you're right in that test.TypeInfo should be being used where possible. The reason it is null in some cases is that the construction of the `MySuite` test suite doesn't pass in any type information like is done in creating the test fixtures. Now I'm not sure if we can say for certain that this TypeInfo will always be present, but if it were the logic could simply be:

>             string containingNamespace = null;
> 
>             if (test.TypeInfo != null)
>             {
>                 containingNamespace = test.TypeInfo.Namespace;
>             }
> 
>             return Match(containingNamespace);

But this means in cases where a TestSuite is constructed without type information then a suite may be skipped when it shouldn't be.

Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: Given the [implementation of `ClassName`](https://github.com/nunit/nunit/blob/3.6/src/NUnitFramework/framework/Internal/Tests/Test.cs#L163-L178), what if you did:

```c#
if (test.TypeInfo != null)
    return Match(test.TypeInfo.Namespace);
else if (test.Method != null)
    return Match(test.Method.MethodInfo.DeclaringType.Namespace)
else
    return false;
```

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs (hunk 3 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: @jnm2 A test suite is represented in the code in many different ways. That representation might be a class, in which case there will be a non-null `TypeInfo` member. But it could represent an assembly or a namespace or a group of methods. In those cases, there is not type to deal with.

NUnit tries hard to separate the concepts of Test, TestSuite etc. from their representation. But it's in the classes derived from Test that the two collide and have to be reconciled.

Some future test implementation might represent a test suite by an XML file or a test case by a class!

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: This should be fixed if we keep it, since a `TestFixture` __is__ a `TestSuite`.

However, the more fundamental question is why we have this branch at all. If there is no class type, then there is no namespace to match, so the test should not be matched.

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                }
+                else
+                {
+                    var outerMostClass = test.ClassName.Substring(0, test.ClassName.IndexOf("+"));
+                    containingNamespace = outerMostClass.Substring(0, outerMostClass.LastIndexOf("."));
+                }
+            }
+
+            return Match(containingNamespace);
+        }
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: It looks as if you are trying to make the `Match` method do the entire job of the filter. But it's only supposed to match those tests that have the namespace you ask for. It's not supposed to find either the antecedents or descendants of those tests. That's up to the parent classes, `ValueMatchFilter` and `TestFilter`. 

Take a look at how `TestNameFilter` does it as an example. It merely matches the name of a test. But when the filter is used, we automatically run all the tests above it in the hierarchy, starting with the assembly, and all the methods below it, if the name for example is a suite.

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs (hunk 1 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: @CharliePoole Is it possible to have a `Method` without a `TypeInfo`? [`Test.ClassName`](https://github.com/nunit/nunit/blob/3.6/src/NUnitFramework/framework/Internal/Tests/Test.cs#L163-L178)

File: src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs (hunk 4 above)
Comment (tom-dudley) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: I think if `ITest` had a `Namespace` property then we could simply just call `Match(test.Namespace)`. Most of the code is really just determining what the namespace of the test actually is, something like:

> var namespace = GetNamespaceOfTest(test);
> Match(namespace);

The reason I was looking at the decedent of the test (in this case a TestSuite) was in order to get the namespace. Perhaps I'm way off the mark with the way I'm approaching this, but I don't see an obvious alternative for knowing what the namespace of a test is.

Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs]: Exactly right. But there's no need to get a namespace unless you are dealing with a test that has a TypeInfo. Anything that doesn't have a TypeInfo should not match at all.

A test without a TypeInfo that is the parent of a matched test will pass the filter and be executed. A test without a TypeInfo that is a child of a matched test will pass the filter and be executed. All this is taken care of by the Pass method. The match method only needs to find tests that match, which means, basically, that  TypeInfo != null and TypeInfo.Namespace == the specified namespace.

**Comment by CharliePoole** [comment]
@tom-dudley Have you made any changes since the last reviews? We don't necessarily get notices unless you post a comment.

**Comment by rprouse** [comment]
There are no commits since Jan 18th. @tom-dudley are you still working on this?

**Comment by tom-dudley** [comment]
Sorry for the delay! I've simplified the `Match` method as @CharliePoole suggested. I hadn't realised about the `return Match(test) || MatchParent(test) || MatchDescendant(test);` in `TestFilter.cs`

File: src/NUnitFramework/tests/Api/TestAssemblyRunnerTests.cs [hunk 5]
```diff
@@ -223,6 +224,21 @@ public void Run_FileNotFound_ReturnsNonRunnableSuite()
...
 
+        [Test]
+        public void RunTestsAction_WithInvalidFilterElement_ThrowsArgumentException()
+        {
+            LoadMockAssembly();
+
+            var ex = Assert.Catch(() =>
+                {
+                    var invalidFilter = TestFilter.FromXml("<filter><invalidElement>foo</invalidElement></filter>");
+                    _runner.Run(this, invalidFilter);
```
Comment (CharliePoole) [inline src/NUnitFramework/tests/Api/TestAssemblyRunnerTests.cs]: No need to run tests. Exception is thrown by TestFilter.FromXml.

File: src/NUnitFramework/tests/Api/TestAssemblyRunnerTests.cs [hunk 6]
```diff
@@ -223,6 +224,21 @@ public void Run_FileNotFound_ReturnsNonRunnableSuite()
...
+                });
+
+            Assert.That(ex, Is.TypeOf<ArgumentException>());
```
Comment (CharliePoole) [inline src/NUnitFramework/tests/Api/TestAssemblyRunnerTests.cs]: You can use `Throws<ArgumentException>` rather than `Catch` and eliminate this line.
